ldb/tx.go

```diff
@@ -346,3 +375,196 @@ func (db *LevelDb) FetchTxBySha(txsha *btcwire.ShaHash) ([]*btcdb.TxListReply, e
 	}
 	return replies, nil
 }
+
+// addrIndexToKey serializes the passed txAddrIndex for storage within the DB.
+// Transactions indexed by address are stored with the following format:
+//   * prefix || hash160 || blockHeight || txoffset || txlen
+// Indexes are stored purely in the key, with blank data for the actual value
+// in order to facilitate ease of iteration by their shared prefix and
+// also to allow limiting the number of returned transactions (RPC).
+// Alternatively, indexes for each address could be stored as an
+// append-only list for the stored value. However, this add unnecessary
+// overhead when storing and retrieving since the *entire* list must
+// be fetched each time.
+func addrIndexToKey(index *txAddrIndex) []byte {
+	record := make([]byte, addrIndexKeyLength, addrIndexKeyLength)
+	copy(record[:2], addrIndexKeyPrefix)
+	copy(record[2:22], index.hash160[:])
+
+	// The index itself.
+	binary.LittleEndian.PutUint64(record[22:30], uint64(index.blkHeight))
+	binary.LittleEndian.PutUint32(record[30:34], uint32(index.txoffset))
+	binary.LittleEndian.PutUint32(record[34:38], uint32(index.txlen))
+
+	return record
+}
+
+// unpackTxIndex deserializes the raw bytes of a address tx index.
+func unpackTxIndex(rawIndex []byte) *txAddrIndex {
+	return &txAddrIndex{
+		blkHeight: int64(binary.LittleEndian.Uint64(rawIndex[0:8])),
+		txoffset:  int(binary.LittleEndian.Uint32(rawIndex[8:12])),
+		txlen:     int(binary.LittleEndian.Uint32(rawIndex[12:16])),
+	}
+}
+
+// BytesPrefix returns key range that satisfy the given prefix.
+// This only applicable for the standard 'bytes comparer'.
+func bytesPrefix(prefix []byte) *util.Range {
+	var limit []byte
+	for i := len(prefix) - 1; i >= 0; i-- {
+		c := prefix[i]
+		if c < 0xff {
+			limit = make([]byte, i+1)
+			copy(limit, prefix)
+			limit[i] = c + 1
+			break
+		}
+	}
+	return &util.Range{Start: prefix, Limit: limit}
+}
+
+// FetchBlockBySha looks up and returns all transactions which either
+// spend from a previously created output of the passed address, or
+// create a new output locked to the passed address. The, `limit` parameter
+// should be the max number of transactions to be returned. Additionally, if the
+// caller wishes to seek forward in the results some amount, the 'seek'
+// represents how many results to skip.
+// NOTE: This function doesn't currently return an error. Instead the length
+// of the TxListReply slice should be checked.
+func (db *LevelDb) FetchTxsForAddr(addr *btcutil.Address, skip int,
+	limit int) ([]*btcdb.TxListReply, error) {
+	db.dbLock.Lock()
+	defer db.dbLock.Unlock()
+
+	scriptAddr := (*addr).ScriptAddress()
+	if len(scriptAddr) != 20 {
```

Comment:
This section concerns me since it's making assumptions about the address and even possible future addresses that may not at all be the same.  I personally think this would be better off type asserting the type of address and using the [Hash160](https://github.com/btcsuite/btcutil/blob/master/address.go#L180) function for `btcutil.AddressPubKeyHash` and `btcutil.AddressScriptHash` and returning an error for unknown address types.

Something like the following:

```Go

var addrKey []byte
switch addr := addr.(type) {
case *btcutil.AddressPubKeyHash:
	hash160 := addr.Hash160()
	addrKey = hash160[:]
case *btcutil.AddressScriptHash:
	hash160 := addr.Hash160()
	addrKey = hash160[:]
case *btcutil.AddressPubKey:
	hash160 := addr.AddressPubKeyHash().Hash160()
	addrKey = hash160[:]
default:
	return nil, ErrUnsupportedAddressType
}
```